src/hessian.jl

```diff
@@ -58,5 +58,17 @@ hessian(f::F, x::SArray, cfg::AllowedHessianConfig{F,H}) where {F,H} = hessian(f
 
 hessian!(result::AbstractArray, f::F, x::SArray) where {F} = jacobian!(result, y -> gradient(f, y), x)
 
-# TODO: Actually implement an optimized version of this. Currently, it just punts to the fallback method.
-hessian!(result::DiffResult, f::F, x::SArray) where {F} = hessian!(result, f, x, HessianConfig(f, result, x))
+hessian!(out::DiffResult, f::F, x::SArray) where {F} = hessian!(out, f, x, HessianConfig(f, out, x))
+
+function hessian!(result::ImmutableDiffResult, f::F, x::SArray) where {F} 
+    d1 = dualize(f, x)
+    d2 = dualize(f, d1)
+    fd2 = f(d2)
+    val = value(value(fd2))
+    grad = extract_gradient(value(fd2), x)
+    hess = extract_jacobian(partials(fd2), x)
+    result = DiffBase.hessian!(result, hess)
+    result = DiffBase.gradient!(result, grad)
+    result = DiffBase.value!(result, val)
+end
+hessian!(result::ImmutableDiffResult, f::F, x::AbstractArray, cfg::AllowedHessianConfig{F,H} = HessianConfig(f, result, x)) where {F,H} = hessian!(result, f, x)
```

Comment:
Yup, that optional argument was the source of my confusion. Also `x::AbstractArray` - I believe that should be `x::SArray`. Otherwise, we'd be throwing away a `cfg` that would need to get reconstructed anyway (e.g. if this method was called with `x::Array`).